python/lsst/ts/eas/m1m3ts_model.py

```diff
@@ -0,0 +1,277 @@
+                while True:
+                    try:
+                        await asyncio.sleep(DORMANT_TIME)
+                    except asyncio.CancelledError:
+                        self.log.info("monitor cancelled")
+                        raise
+
+    async def apply_setpoints(
+        self, *, m1m3ts_remote: salobj.Remote, setpoint: float
+    ) -> None:
+        glycol_setpoint = setpoint + self.glycol_setpoint_delta
+        heaters_setpoint = setpoint + self.heater_setpoint_delta
+        self.log.debug(f"Setting MTM1MTS: {glycol_setpoint=} {heaters_setpoint=}")
+        if hasattr(m1m3ts_remote, "cmd_applySetpoints"):
+            await m1m3ts_remote.cmd_applySetpoints.set_start(
+                glycolSetpoint=glycol_setpoint,
+                heatersSetpoint=heaters_setpoint,
+            )
+        else:
+            await m1m3ts_remote.cmd_applySetpoint.set_start(
+                glycolSetpoint=glycol_setpoint,
+                heatersSetpoint=heaters_setpoint,
+            )
+
+    async def follow_ess_indoor(self, *, m1m3ts_remote: salobj.Remote) -> None:
+        self.log.debug("follow_ess_indoor")
+
+        async with salobj.Remote(
+            domain=self.domain,
+            name="ESS",
+            index=self.indoor_ess_index,
+            include=("temperature",),
+        ) as ess_remote:
+            while True:
+                if self.dome_model.dome_is_closed:
+                    await asyncio.sleep(DORMANT_TIME)
+                    continue
+
+                temperatures = []
+                start_time = utils.current_tai()
+
+                # Collect average indoor ESS temperature for
+                # m1m3_setpoint_cadence seconds.
+                while utils.current_tai() - start_time < self.m1m3_setpoint_cadence:
+                    try:
+                        sample = await ess_remote.tel_temperature.aget(timeout=10)
+                        await asyncio.sleep(0)  # Make sure we get CancelledError
+                        temperatures.append(sample.temperatureItem[0])
+                    except asyncio.CancelledError:
+                        raise
+                    except Exception as e:
+                        self.log.warning(
+                            f"Failed to get ESS:{self.indoor_ess_index} temperature: {e!r}"
+                        )
+
+                if not temperatures:
+                    self.log.error(
+                        "No temperature samples were collected. CSC will fault."
+                    )
+                    raise RuntimeError("No temperature samples were collected.")
+
+                average_temperature = sum(temperatures) / len(temperatures)
+                self.log.debug(
+                    f"Collected {len(temperatures)} ESS:{self.indoor_ess_index}"
+                    f" samples with {average_temperature=}."
+                )
+
+                if average_temperature is None or math.isnan(average_temperature):
```

Comment:
I think this shall not occur.. or at least the temperatures samples shall be tested if their aren't nan right when being received, and timer restarted if nan is received.